sonar-batch/src/main/java/org/sonar/batch/issue/ModuleIssues.java

```diff
@@ -37,55 +46,126 @@
+
+    Rule rule = validateRule(issue);
+    ActiveRule activeRule = activeRules.find(issue.ruleKey());
     if (activeRule == null) {
       // rule does not exist or is not enabled -> ignore the issue
       return false;
     }
-    updateIssue(issue, rule, activeRule);
-    if (filters.accept(issue)) {
-      cache.put(issue);
+
+    String primaryMessage = Strings.isNullOrEmpty(issue.primaryLocation().message()) ? rule.name() : issue.primaryLocation().message();
+    org.sonar.api.batch.rule.Severity overriddenSeverity = issue.overriddenSeverity();
+    Severity severity = overriddenSeverity != null ? Severity.valueOf(overriddenSeverity.name()) : Severity.valueOf(activeRule.severity());
+
+    builder.clear();
+    locationBuilder.clear();
+    // non-null fields
+    builder.setSeverity(severity);
+    builder.setRuleRepository(issue.ruleKey().repository());
+    builder.setRuleKey(issue.ruleKey().rule());
+    builder.setAttributes(KeyValueFormat.format(issue.attributes()));
+    builder.setMsg(primaryMessage);
+    locationBuilder.setMsg(primaryMessage);
+
+    locationBuilder.setComponentRef(component.batchId());
+    TextRange primaryTextRange = issue.primaryLocation().textRange();
+    applyTextRange(primaryTextRange);
+    if (primaryTextRange != null) {
+      builder.setLine(primaryTextRange.start().line());
+    }
+    builder.setPrimaryLocation(locationBuilder.build());
+    Double effortToFix = issue.effortToFix();
+    if (effortToFix != null) {
+      builder.setEffortToFix(effortToFix);
+    }
+    applyAdditionalLocations(issue);
+    applyExecutionFlows(issue);
+    BatchReport.Issue rawIssue = builder.build();
+
+    if (filters.accept(key, rawIssue)) {
+      write(component, rawIssue);
       return true;
     }
     return false;
   }
 
-  private static void validateRule(DefaultIssue issue, @Nullable Rule rule) {
+  private void applyAdditionalLocations(Issue issue) {
+    for (org.sonar.api.batch.sensor.issue.IssueLocation additionalLocation : issue.locations()) {
+      locationBuilder.clear();
+      String locationComponentKey = ((DefaultInputComponent) additionalLocation.inputComponent()).key();
+      locationBuilder.setComponentRef(componentCache.get(locationComponentKey).batchId());
+      String message = additionalLocation.message();
+      if (message != null) {
+        locationBuilder.setMsg(message);
+      }
+      applyTextRange(additionalLocation.textRange());
+      builder.addAdditionalLocation(locationBuilder.build());
+    }
+  }
+
+  private void applyExecutionFlows(Issue issue) {
+    for (ExecutionFlow executionFlow : issue.executionFlows()) {
+      flowBuilder.clear();
+      for (org.sonar.api.batch.sensor.issue.IssueLocation location : executionFlow.locations()) {
+        locationBuilder.clear();
+        String locationComponentKey = ((DefaultInputComponent) location.inputComponent()).key();
+        locationBuilder.setComponentRef(componentCache.get(locationComponentKey).batchId());
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) NullPointerException might be thrown as 'get' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2259)